From 63dbf6b3b9e6e781df1a6a64e609b10e23969681 Mon Sep 17 00:00:00 2001
From: Paul Eggert <eggert@cs.ucla.edu>
Date: Wed, 15 Apr 2026 12:00:17 -0700
Subject: gzip: don’t mishandle .lzh after .Z
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Problem reported by Michał Majchrowicz.
* unlzh.c (read_c_len): Clear left and right when n == 0.
---
 NEWS    | 4 ++++
 THANKS  | 1 +
 unlzh.c | 6 ++++++
 3 files changed, 11 insertions(+)

diff --git a/NEWS b/NEWS
index 45ac0af..c053c57 100644
--- a/NEWS
+++ b/NEWS
@@ -11,6 +11,10 @@ GNU gzip NEWS                                    -*- outline -*-
   A use of uninitialized memory on some malformed inputs has been fixed.
   [bug present since the beginning]
 
+  A buffer overflow has been fixed when decompressing an .lzh file
+  after decompressing a .Z file.
+  [bug present since the beginning]
+
 ** Changes in behavior
 
   gzip -l now reports "-Inf%" instead of "0.0%" for the infinite
diff --git a/THANKS b/THANKS
index 4e545d9..a7d25e4 100644
--- a/THANKS
+++ b/THANKS
@@ -187,6 +187,7 @@ Richard Lloyd           R.K.Lloyd@csc.liv.ac.uk
 Mohamed Maatallah	zephyrofficialdiscord@gmail.com
 David J. MacKenzie	djm@eng.umd.edu
 John R MacMillan        john@chance.gts.org
+Michał Majchrowicz	mmajchrowicz@afine.com
 Ron Male                male@eso.mc.xerox.com
 Jakub Martisko		jamartis@redhat.com
 Don R. Maszle           maze@bea.lbl.gov
diff --git a/unlzh.c b/unlzh.c
index 3320196..a6cf109 100644
--- a/unlzh.c
+++ b/unlzh.c
@@ -232,6 +232,12 @@ read_c_len ()
         c = getbits(CBIT);
         for (i = 0; i < NC; i++) c_len[i] = 0;
         for (i = 0; i < 4096; i++) c_table[i] = c;
+
+        /* Needed in case LEFT and RIGHT are reused from a previous
+           LZW decompression.  It may be overkill to clear all of both
+           arrays, but nobody has had time to analyze this carefully.  */
+        memzero(left, (2 * NC - 1) * sizeof *left);
+        memzero(right, (2 * NC - 1) * sizeof *left);
     } else {
         i = 0;
         while (i < n) {
-- 
cgit v1.3

